manager/common/license.go

```diff
@@ -0,0 +1,175 @@
+    counter map[string]int // current number of licenses used per vendor
+	clientMutex sync.Mutex
+	clients map[string]*Client // client connection metadata
+    total map[string]int // total licenses available per vendor
+	queue []string
+}
+
+func (c *LicenseCounter) increment(vendor string, num int, hash string) int {
+    c.counterMutex.Lock()
+    if c.counter[vendor] < c.total[vendor] {
+		log.Printf("License acquired by %s \n", hash)
+        c.counter[vendor] += num
+    } else {
+		log.Fatalf("Cannot acquire more licenses than available \n")
+	}
+    defer c.counterMutex.Unlock()
+    return c.counter[vendor]
+}
+
+func (c *LicenseCounter) decrement(vendor string, num int, hash string) int {
+	c.counterMutex.Lock()
+	if c.counter[vendor] - num >= 0 {
+		log.Printf("License released by %s \n", hash)
+		c.counter[vendor] -= num
+	} else {
+		log.Fatalf("Cannot release a negative number of licenses \n")
+	}
+	defer c.counterMutex.Unlock()
+	return c.counter[vendor]
+}
+
+func (c *LicenseCounter) enqueue(key string, val *Client) int {
+	c.clientMutex.Lock()
+	c.queue = append(c.queue, key)
+	c.clients[key] = val
+	defer c.clientMutex.Unlock()
+	return len(c.queue)
+}
+
+func (c *LicenseCounter) dequeue(key string) {
+	c.clientMutex.Lock()
+	c.queue = c.queue[1:]
+	delete(c.clients, key)
+	defer c.clientMutex.Unlock()
+}
+
+func (c *LicenseCounter) del(key string) int {
+	index := -1
+	c.clientMutex.Lock()
+	for i := 0; i < len(c.queue); i++ {
+		if c.queue[i] == key {
+			c.queue = append(c.queue[:i], c.queue[i+1:]...)
+			delete(c.clients, key)
+			log.Printf("Client %s removed from queue \n", key)
+			index = i
+			break
+		}
+	}
+	if index == -1 {
+		log.Fatalf("%s is not in queue \n", key)
+	}
+	defer c.clientMutex.Unlock()
+	return index
+}
+
+type Client struct {
+	IP string
+	User string
+	Start time.Time
+}
+
+var clients = make(map[string]*Client)
+var totalCadenceLic = 3
+var totalXilinxLic = 3
+var licenseCounter = LicenseCounter{counter: map[string]int{"xilinx": 0, "cadence": 0},
+									total: map[string]int{"xilinx": totalXilinxLic, "cadence": totalCadenceLic},
+									clients: make(map[string]*Client),
+									queue: make([]string, 0)}
+type Server struct {
+}
+
+func (s *Server) KeepAlive(stream rpc_license.License_KeepAliveServer) error {
+	var status error
+	var hash string
+	var vendor string
+	var feature string
+	for {
+		recv, err := stream.Recv()
+		if err == io.EOF {
+			status = nil
+			break
+		}
+		if err != nil {
+			status = err
+			break
+		}
+		hash, vendor, feature = recv.Hash, recv.Vendor, recv.Feature
+		err = stream.Send(&rpc_license.KeepAliveMessage{Hash: hash, Vendor: vendor, Feature: feature})
+		if err != nil {
+			status = err
+			break
+		}
+	}
+	inUse := licenseCounter.decrement(vendor, 1, hash)
+	log.Printf("%s %s licenses in use: %d/%d \n", vendor, feature, inUse, licenseCounter.total[vendor])
+	return status
+}
+
+func (s *Server) Polling(stream rpc_license.License_PollingServer) error {
```

Comment:
It took me some time to understand how the code works, there are parts that are not clear to me.

Correct me if wrong:
- client opens a Polling stream. On this stream, the client sends PollingRequests every few seconds. If the requested license is available, then acquired: true is returned, and the stream is closed. Otherwise, acquired: false is returned.
- client opens a KeepAlive stream. When this stream is closed, the license is released.

The keepalive messages exchanged in this stream... serve no purpose? as there is no timeout? it's only based on the Read() returning an error?

Also, the hash is ... not necessary as is, right? Eg, given that a stream must be kept open, the state may as well be tied to the stream, without a hash? in the queue you could put a client object or the stream itself, right?

In my previous comment, I was recommending either using streams, *or* passing around a hash.

My  take:
- we only need a hash, a cookie/token exchanged between client and server, if we either don't use streaming, eg, simple rpc messages, stateless, or if we want to support re-connecting if the stream fails/is closed.
- don't fully understand why we need two streams here.

We could get away with just one, right? Which would also make the code resilient to a client that crashes/never sends a keepalive? And we could remove the hash entirely if we use streaming?

(well, unless we add a bit of code to re-connect without losing a license, which could be valuable)

Reply:
Ok, I can consolidate my code to just have a single bidirectional stream. I do not want to implement any reconnect logic because sometimes disconnections are intentional like when a user on the client side does `CTRL+C` when the bazel test/build is in progress but the user wants to abort.